File: modules/kscan-live-vto-native/ios/Core/LiveVtoVec2.swift
import Foundation

/// A plain 2D float vector.
///
/// Field-for-field port of Android's `LiveVtoVec2.kt` (`Vec2`), itself
/// deliberately not `android.graphics.PointF` for the same reason this is
/// deliberately not `CGPoint`: every geometry stage of this runtime --
/// anchors, control-point targets, rigid fit, rigid gate, mesh deformation --
/// runs on this type and therefore has zero UIKit/ExpoModulesCore
/// dependencies, which is what makes the whole geometry pipeline runnable in
/// a plain SwiftPM host test (`swift test`), independent of a simulator or
/// device, and structurally incapable of touching a `UIView` or the main
/// thread.
///
/// Values are immutable; every operation returns a new instance.
public struct Vec2: Equatable, Hashable {
  public let x: Float
  public let y: Float

  public init(_ x: Float, _ y: Float) {
    self.x = x
    self.y = y
  }

  public static func + (a: Vec2, b: Vec2) -> Vec2 { Vec2(a.x + b.x, a.y + b.y) }
  public static func - (a: Vec2, b: Vec2) -> Vec2 { Vec2(a.x - b.x, a.y - b.y) }
  public static func * (a: Vec2, s: Float) -> Vec2 { Vec2(a.x * s, a.y * s) }

  public func length() -> Float { Float(Foundation.hypot(Double(x), Double(y))) }

  public func normalized() -> Vec2 {
    let l = length()
    return l < 1e-6 ? Vec2(0, 0) : Vec2(x / l, y / l)
  }

  /// True when either component is NaN or infinite -- the finite-ness guard every stage boundary checks.
  public var isFinite: Bool { x.isFinite && y.isFinite }
}

public func dot(_ a: Vec2, _ b: Vec2) -> Float { a.x * b.x + a.y * b.y }

File: modules/kscan-live-vto-native/android/src/main/java/expo/modules/kscanlivevtonative/LiveVtoVec2.kt
package expo.modules.kscanlivevtonative

import kotlin.math.hypot

/**
 * A plain 2D float vector.
 *
 * Deliberately NOT `android.graphics.PointF`. Every geometry stage of this
 * runtime -- anchors, control-point targets, rigid fit, rigid gate, mesh
 * deformation -- runs on this type and therefore has zero Android
 * dependencies, which buys two things the mission requires directly:
 *
 *  1. The whole geometry pipeline runs in a plain JVM unit test
 *     (`:kscan-live-vto-native:testDebugUnitTest`), so cross-runtime
 *     conformance against the P3-A reference oracle is measured
 *     numerically, deterministically, and repeatably -- not read off a
 *     screenshot and not gated on an emulator being healthy
 *     (amendment D8: "numerical geometry evidence is authoritative for
 *     geometry; screenshots are supplementary").
 *
 *  2. Deformation compute is structurally incapable of touching a Canvas,
 *     a View, or the UI thread (amendment D10) -- it consumes a BodyFrame
 *     and produces an immutable snapshot, and the only Android type in the
 *     render path lives in LiveVtoTestRenderView's draw call.
 *
 * Values are immutable; every operation returns a new instance.
 */
data class Vec2(val x: Float, val y: Float) {
  operator fun plus(o: Vec2) = Vec2(x + o.x, y + o.y)
  operator fun minus(o: Vec2) = Vec2(x - o.x, y - o.y)
  operator fun times(s: Float) = Vec2(x * s, y * s)
  fun length(): Float = hypot(x.toDouble(), y.toDouble()).toFloat()
  fun normalized(): Vec2 { val l = length(); return if (l < 1e-6f) Vec2(0f, 0f) else Vec2(x / l, y / l) }

  /** True when either component is NaN or infinite -- the finite-ness guard every stage boundary checks. */
  val isFinite: Boolean get() = x.isFinite() && y.isFinite()
}

fun dot(a: Vec2, b: Vec2): Float = a.x * b.x + a.y * b.y
